rust-runtime/aws-smithy-http-server/src/routing/into_make_lambda_service.rs

```diff
@@ -0,0 +1,203 @@
+        self.service.poll_ready(cx).map_err(|err| err.into())
+    }
+
+    /// Lambda handler function
+    /// Parse Lambda request as hyper request,
+    /// serialize hyper response to Lambda response
+    fn call(&mut self, event: Request) -> Self::Future {
+        // As recommended in https://github.com/tower-rs/tower/issues/547
+        let clone = self.service.clone();
+        let mut inner = std::mem::replace(&mut self.service, clone);
+
+        let fut = async move {
+            // Parse request
+            let hyper_request = lambda_to_hyper_request(event)?;
+
+            // Call Hyper service when request parsing succeeded
+            let response = inner.call(hyper_request).await?;
+
+            // Return Lambda response after finished calling inner service
+            hyper_to_lambda_response(response).await
+        };
+        MakeRouteLambdaServiceFuture::new(Box::pin(fut))
+    }
+}
+
+opaque_future! {
+    /// Response future for [`IntoMakeLambdaService`] services.
+    pub type MakeRouteLambdaServiceFuture = BoxFuture<'static, Result<Response<LambdaBody>, BoxError>>;
+}
+
+fn lambda_to_hyper_request(request: Request) -> Result<HyperRequest, BoxError> {
+    tracing::debug!("Converting Lambda to Hyper request...");
+    // Raw HTTP path without any stage information
+    let raw_path = request.raw_http_path();
+    let (mut parts, body) = request.into_parts();
+    let mut path = String::from(parts.uri.path());
+    if !raw_path.is_empty() && raw_path != path {
+        tracing::debug!("Recreating URI from raw HTTP path.");
+        path = raw_path;
+        let uri_parts: uri::Parts = parts.uri.into();
+        let path_and_query = uri_parts.path_and_query.unwrap();
+        if let Some(query) = path_and_query.query() {
+            path.push('?');
+            path.push_str(query);
+        }
+        parts.uri = uri::Uri::builder()
+            .authority(uri_parts.authority.unwrap())
+            .scheme(uri_parts.scheme.unwrap())
+            .path_and_query(path)
+            .build()
+            .unwrap();
+    }
+    let body = match body {
+        LambdaBody::Empty => HyperBody::empty(),
+        LambdaBody::Text(s) => HyperBody::from(s),
+        LambdaBody::Binary(v) => HyperBody::from(v),
+    };
+    // We need to maintain all parts including the extensions
+    let req = http::Request::from_parts(parts, body);
+    tracing::debug!("Hyper request converted successfully.");
+    Ok(req)
+}
+
+async fn hyper_to_lambda_response<B>(response: Response<B>) -> Result<Response<LambdaBody>, BoxError>
+where
+    B: Body + Debug,
+    <B as Body>::Error: Error + Send + Sync + 'static,
+{
+    tracing::debug!("Converting Hyper to Lambda response...");
+    // Divide response into headers and body
+    let (parts, body) = response.into_parts();
+    let body = hyper::body::to_bytes(body).await?;
+    let res = Response::from_parts(parts, LambdaBody::from(body.as_ref()));
```

Comment:
I deployed a Lambda with the Pokémon service, always converting the Smithy response into a `lambda_http::Body::Binary` variant like here, and I got binary output in the payload, even when setting `Accept: application/json`.

For example, I got `eyJjYWxsc19jb3VudCI6MH0=` for the `/stats` endpoint, which base64 decodes to the otherwise expected response, `{"calls_count":0}`.

---

@eduardomourar I had presumed that you were working with this patch applied in your Lambdas. How are you working around this for smithy-rs operation outputs with text payloads? My thinking right now is that it's impossible to provide a Lambda response without model-specific knowledge that tells you whether what you're returning is a text or a binary payload.

Reply:
@eduardomourar I created a reproducer repo and explained the problem in more detail here: https://github.com/awslabs/aws-lambda-rust-runtime/pull/491#discussion_r903957717.

There's a difference in how API Gateway returns the response depending on if the API is HTTP or REST. HTTP APIs work fine, so maybe that is the reason why you did not experience this erroneous behavior using this patch.

Reply:
We (Hugo and I) investigated our configuration and we are using the REST API. Our current deployment is still using a previous version of this branch (while using the latest released version of the `lambda_http` crate) and it gives the expected response. Hugo is currently working to update to use both changes from this PR as well [the other PR from lambda runtime](https://github.com/awslabs/aws-lambda-rust-runtime/pull/491). Most update should happen in that other PR.

Reply:
> Most update should happen in that other PR.

Yeah, I believe after that PR lands, the only thing that we would need to do here is the "stripping out the stage from the request URI path" part right?

Reply:
Yes, hopefully so. Some extra logic might be placed in this repo. We will see.